python/cudf/cudf/core/indexed_frame.py

```diff
@@ -445,6 +445,70 @@ def sort_index(
             out = out.reset_index(drop=True)
         return self._mimic_inplace(out, inplace=inplace)
 
+    def hash_values(self, method="murmur3"):
+        """Compute the hash of values in this column.
+
+        Parameters
+        ----------
+        method : {'murmur3', 'md5'}, default 'murmur3'
+            Hash function to use:
+            * murmur3: MurmurHash3 hash function.
+            * md5: MD5 hash function.
+
+        Returns
+        -------
+        Series
+            A Series with hash values.
+
+        Examples
+        --------
+        **Series**
+
+        >>> import cudf
+        >>> series = cudf.Series([10, 120, 30])
+        >>> series
+        0     10
+        1    120
+        2     30
+        dtype: int64
+        >>> series.hash_values(method="murmur3")
+        0   -1930516747
+        1     422619251
+        2    -941520876
+        dtype: int32
+        >>> series.hash_values(method="md5")
+        0    7be4bbacbfdb05fb3044e36c22b41e8b
+        1    947ca8d2c5f0f27437f156cfbfab0969
+        2    d0580ef52d27c043c8e341fd5039b166
+        dtype: object
+
+        **DataFrame**
+
+        >>> import cudf
+        >>> df = cudf.DataFrame({"a": [10, 120, 30], "b": [0.0, 0.25, 0.50]})
+        >>> df
+             a     b
+        0   10  0.00
+        1  120  0.25
+        2   30  0.50
+        >>> df.hash_values(method="murmur3")
+        0    -330519225
+        1    -397962448
+        2   -1345834934
+        dtype: int32
+        >>> df.hash_values(method="md5")
+        0    57ce879751b5169c525907d5c563fae1
+        1    948d6221a7c4963d4be411bcead7e32b
+        2    fe061786ea286a515b772d91b0dfcd70
+        dtype: object
+        """
+        # Note that both Series and DataFrame return Series objects from this
+        # calculation, necessitating the unfortunate circular reference to the
+        # child class here.
```

Comment:
Is this somehow worse than the previous code referring to `Series` within the `Series.hash_values`? It doesn't seem like a problem to me - the comment block might be unnecessary.

Reply:
Yes, because Series inherits from IndexedFrame. This is code in a parent class that relies on knowing a specific child exists.

Reply:
Are you saying this is circular (or awkward) from the perspective of Python imports, or from a perspective of object-oriented purity? This implementation looks correct and feels only slightly awkward to me. This could be phrased as, "A `_ThingContainer` has derived classes `SingleThingContainer` and `MultiThingContainer`. A `_ThingContainer` has a method `collapseThings` that returns a `SingleThingContainer`, which can be called by its derived classes," with which I have no qualms.

Regardless, I don't think there is a good way around it. If this were indicative of a design flaw that we could fix, then I might say we should keep the comment (as a sort of TODO that indicates the circularity should be removed with some improved design). However, I hesitate to call this "unfortunate" if it is _necessary_.

(This is not a big deal - happy to close the conversation and let you resolve however you wish. Just wanted to clarify my thoughts.)

Reply:
Interesting viewpoint. What you're describing sounds more like a choice of how to leverage a _type system_ (e.g. principles of [covariance / contravariance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science))) with the class hierarchy and methods' return types.

Here's why I think this is fine to keep in `IndexedFrame`, if you wish:
- Many methods of `IndexedFrame` (or its parent `Frame`) have _covariant_ return types and return an instance of `self.__class__` (e.g. [round](https://github.com/rapidsai/cudf/blob/194e209d9c7f4ac4271389738df6db3fd067f516/python/cudf/cudf/core/indexed_frame.py#L992))
- The return type of `hash_values` is _invariant_ to the choice of subclass. Other invariant methods like `Frame.__len__` can return an `int` or a `str` or any type that doesn't depend on the choice of subclass. It just happens that the invariant return type of `IndexedFrame.hash_values` is `Series`, a child class of `IndexedFrame`.

It is partly a question of what semantics we choose to adhere to in `IndexedFrame`. Regardless, not all methods of `IndexedFrame` have covariant return types (and it would be silly to require that), so I'm not sure what grounds would justify removing this from `IndexedFrame`, especially since the underlying [Cython hash](https://github.com/rapidsai/cudf/blob/194e209d9c7f4ac4271389738df6db3fd067f516/python/cudf/cudf/_lib/hash.pyx#L57) doesn't care. Methods with invariant return types are perfectly fine to include, and I think this is one of them.

Reply:
I think it's fine to keep here. I don't think the following is necessarily true:

```
IndexedFrame.hash_values is actually semantically different for different subclasses of IndexedFrame.
```

`hash_values` hashes each row of a Frame. A `Series` is viewed as `Frame` with `N` rows, rather than a `Frame` with `N` columns. I would consider it surprising behaviour if `Series.hash_values` instead returned a scalar.